```{r}
library(Seurat)
library(edgeR)
library(dplyr)
```

```{r}
# Load SO
load("../data/SO_Gli_Ascl_comb_int.Robj")
```

```{r}
# Create edgeR object
edge_obj = DGEList(counts = as.matrix(pbmc.combined@assays$RNA@counts), samples = pbmc.combined@meta.data)
```

```{r}
# Remove mithocondrial genes
mit_genes = grep("^mt-", rownames(edge_obj))
edge_obj = edge_obj[-mit_genes,]
```

```{r}
# Get gene annotation + gene length for further lst normalization
# Use asia ensembl since european is under maintenance

lst = list()
lst$genes = edge_obj$counts %>% rownames
lst$attributes = c("mgi_symbol" , "start_position", "end_position")

lst$bmart = biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL" , dataset = "mmusculus_gene_ensembl", 
            host = "asia.ensembl.org")

feature_info = biomaRt::getBM(attributes = lst$attributes, filters = "mgi_symbol", 
        values = lst$genes, mart = lst$bmart)

#keep genes which are not common to initial edge object and biomart result 

x = match(lst$genes, feature_info$mgi_symbol)
feature_info = feature_info[x, ]
feature_info = na.omit(feature_info)

gene_meta_data = data.frame(id = feature_info$mgi_symbol, 
                            start_position = feature_info$start_position,
                            end_position = feature_info$end_position,
                            length = feature_info$end_position - feature_info$start_position)
rownames(gene_meta_data) = gene_meta_data$id

## Generating TPM counts
edge_obj$counts = edge_obj$counts[feature_info$mgi_symbol,]
x = edge_obj$counts / gene_meta_data[feature_info$mgi_symbol,'length']

tpm = t(t(x) * 1e6 / colSums(x))
```

```{r}
# Filtering genes with TMP > 1 and genes which are present in more than 25% of cells
# as in https://www.nature.com/articles/nmeth.4612. TPMs because of the sequencing length bias https://f1000research.com/articles/6-595.
keep = rowSums(tpm > 1) >= 0.25*ncol(edge_obj$counts)
table(keep)
edge_obj = edge_obj[keep,]
```

```{r}
batch <- edge_obj$samples$batch_group
genotype <- edge_obj$samples$type
cluster <- edge_obj$samples$cluster
```

```{r}
# Create 2 batches for design
alt_batch =as.character(batch)
alt_batch[alt_batch == 'Ascl1_12wk_1'] ='B1'
alt_batch[alt_batch == 'Ascl1_12wk_2'] ='B2'
alt_batch[alt_batch == 'Ascl1_5day_1'] ='B1'
alt_batch[alt_batch == 'Ascl1_5day_2'] ='B2'
alt_batch[alt_batch == 'Gli1_12wk']    ='B1'
alt_batch[alt_batch == 'Gli1_5day_1']  ='B1'
alt_batch[alt_batch == 'Gli1_5day_2']  ='B1'
alt_batch =as.factor(alt_batch)
table(alt_batch)

design =model.matrix( ~0 + genotype + alt_batch)
```

```{r}
# norm factors
edge_obj <- calcNormFactors(edge_obj)
# Perform generalized log-linear model with quasi-likelihood tests
edge_obj <- estimateDisp(edge_obj, design)
fit <- glmQLFit(edge_obj, design)
```

```{r}
# multidimensional scaling plot that shows distances between samples
limma::plotMDS(edge_obj, col = as.numeric(as.factor(edge_obj$samples$type)), pch = 19)
```

```{r}
# Plot gene-wise biological coefficient of variantion + quasi-likelihood dispersion
plotBCV(edge_obj)
plotQLDisp(fit)
```

```{r}
                    ## Comparing Gli and Ascl genotypes with glmQLF
avsgli <- makeContrasts(genotypeAscl1-genotypeGli1, levels=design)
qlf_genot <- glmQLFTest(fit, contrast=avsgli)
```

```{r}
tt_genot <- topTags(qlf_genot, adjust.method = 'BH', sort.by = 'PValue',
                    n = 1e6, p.value = 0.05)
tt_genot_df <- data.frame(gene = rownames(tt_genot),
                       tt_genot)
tt_genot_df$analysis <- 'ascl_vs_gli_glmqlf'
```

```{r}
hist(tt_genot$table$PValue, 50)
hist(tt_genot$table$FDR, 50)
plotSmear(qlf_genot)
```

```{r}
# Heat map of top 50 differentially expressed genes
logCPM <- cpm(dge, prior.count=2, log=TRUE)
rownames(logCPM) <- rownames(dge$counts)
colnames(logCPM) <- rownames(dge$samples)
## table is already sorted
o <- head(rownames(tt_genot$table), 50)
logCPM <- logCPM[o,]
logCPM <- t(scale(t(logCPM)))
col.pan <- gplots::colorpanel(100, "blue", "white", "red")
## sampling only 100 cells
set.seed(1)
logCPM <- logCPM[,sample(1:ncol(logCPM), 100)]
pheatmap::pheatmap(logCPM, color = col.pan,
         annotation_col = dge$samples[,c('cluster', 'batch_group', 'type')])
```

                    Use Logistic regression
In this case, we require 20% of the cells in either Ascl1 or Gli1 to express the gene.
Note about differential expression results: it hasthe following columns,
- gene: the gene symbol
- p_val : p_val (unadjusted)
- avg_logFC : log fold-chage of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.
- pct.1 : The percentage of cells where the feature is detected in the first group
- pct.2 : The percentage of cells where the feature is detected in the second group
- p_val_adj : Adjusted p-value, based on bonferroni correction using all features in the dataset.
- analysis: a tag describing the analysis carried out

```{r}
                    # Compute DEG betwee Gli vs Ascl independently of clusters
Idents(pbmc.combined) = "orig.ident"
pbmc.combined@meta.data$alt_batch = alt_batch

# Compute genes
markers_all = FindMarkers(pbmc.combined,
                       ident.1 =  'Ascl1',
                       ident.2 = 'Gli1',
                       test.use = "LR",
                       latent.vars = "alt_batch",
                       min.pct = 0.2)

markers_all[order(markers_all$p_val_adj),] %>% filter(p_val < 0.05)
markers_all = data.frame(gene = rownames(markers_all),
                      markers_all)
```

```{r}
                    ## DEG for Ascl vs Gli for cluster ndNSC
DefaultAssay(pbmc.combined) = "RNA"
Idents(pbmc.combined) = 'combi'
markers_ndNSC = FindMarkers(pbmc.combined,
                       ident.1 =  'Ascl1_ndNSC',
                       ident.2 = 'Gli1_ndNSC',
                       test.use = "LR",
                       latent.vars = "alt_batch",
                       min.pct = 0.2)
markers_ndNSC[order(markers_ndNSC$p_val_adj),] %>% filter(p_val < 0.05)
markers_ndNSC = data.frame(gene = rownames(markers_ndNSC),
                      markers_ndNSC)
```

```{r}
                    ## DEG for Ascl vs Gli for cluster dNSC
DefaultAssay(pbmc.combined) = "RNA"
Idents(pbmc.combined) = 'combi'
markers_dNSC = FindMarkers(pbmc.combined,
                       ident.1 =  'Ascl1_dNSC',
                       ident.2 = 'Gli1_dNSC',
                       test.use = "LR",
                       latent.vars = "alt_batch",
                       min.pct = 0.2)
markers_dNSC[order(markers_dNSC$p_val_adj),] %>% filter(p_val < 0.05)
markers_dNSC = data.frame(gene = rownames(markers_dNSC),
                      markers_dNSC)
```
